File: src/nishikihebi/chat/session.py
from __future__ import annotations

from typing import Protocol
from uuid import uuid4

from langchain_core.messages import HumanMessage
from langgraph.graph.state import CompiledStateGraph


class Session(Protocol):
    def ask(self, question: str) -> str: ...


class ChatSession:
    def __init__(self, graph: CompiledStateGraph, thread_id: str) -> None:
        self.graph = graph
        self.thread_id = thread_id

    def ask(self, question: str) -> str:
        result = self.graph.invoke(
            {"messages": [HumanMessage(content=question)]},
            config={"configurable": {"thread_id": self.thread_id}},
        )
        return result["messages"][-1].content


def start_session(graph: CompiledStateGraph) -> ChatSession:
    return ChatSession(graph, thread_id=str(uuid4()))

File: apps/server/agents/chat/session.ts
import { randomUUID } from "node:crypto";
import { HumanMessage } from "@langchain/core/messages";
import type { ChatGraph } from "./graph.js";

export interface Session {
  ask(question: string): Promise<string>;
}

export class ChatSession implements Session {
  constructor(
    private readonly graph: ChatGraph,
    readonly threadId: string,
  ) {}

  async ask(question: string): Promise<string> {
    const result = await this.graph.invoke(
      { messages: [new HumanMessage(question)] },
      { configurable: { thread_id: this.threadId } },
    );
    const reply = result.messages.at(-1);
    return typeof reply?.content === "string" ? reply.content : "";
  }
}

export function startSession(graph: ChatGraph): ChatSession {
  return new ChatSession(graph, randomUUID());
}
